### ISSUE TITLE
[Forms][Android][iOS] Notification Player

### ISSUE THREAD

**Issue by valentin-debris** [martijn00/XamarinMediaManager#439]
## 🔙 Regression

Since the release of the version 0.5.* (Thansk for that by the way !), I noticed a regression with the Notification which displays the controls.

### Old (and correct) behavior
In iOS, it was visible
In Android (API >= 26) it wasn't working due to the missing channel

### Current behavior
In iOS, I don't see it anymore, even on lock screen.
In Android, it works now but I can't remove it. If I click on stop, it stop the song but it's still displayed. I have to Force kill the app to remove it.

Also, it's not a bug but I want to know if it's possible to remove somes controls in the notification, because the audio will be a live streaming, and I don't want to show the prev-next-pause button.

### Reproduction steps
If it's not reproductible, I will try to create a dummy repo for that.

### Configuration

**Version:** 0.5.3

**Platform:** 
- [x] :iphone: iOS
- [x] :robot: Android
- [x] :monkey: Xamarin.Forms

**Comment by martijn00** [comment]
I think this is a regression on iOS indeed.
On Android this is now controlled by Exoplayer. You can easily access that and change the behavior, also see: https://medium.com/google-exoplayer/the-mediasession-extension-for-exoplayer-82b9619deb2d

I see the use case to make this more convenient, but not sure when i have time to look at it. A PR would be appreciated.

**Comment by valentin-debris** [comment]
Hello @martijn00 ,

I'm trying to change the behavior using Exoplayer but I don't know where to start !...
I know I have to override the behavior of MediaManager.Platforms.Android.MediaSession.MediaBrowserService, specially for the MediaQueue_CollectionChanged method.

I tried adding in MainActivity.cs the following line after the init() but because PlayerNotificationManager is protected, it doesn't work.
`MediaManager.Platforms.Android.MediaSession.MediaBrowserService.PlayerNotificationManager.SetUseNavigationActions(false);`

If you can give me quick help on where to start, I would appreciate.

Thanks !

**Comment by martijn00** [comment]
Currently you need to derive from the normal class and override it. I think i can open up some classes to be public visible. Maybe also add a property to enable or disable notification controls. I'll try and add it sometime soon.

**Comment by martijn00** [comment]
@valentin-debris can you test this:

CrossMediaManager.Current.NotificationManager.ShowNavigationControls = false;

Available in 0.5.6.

**Comment by valentin-debris** [comment]
@martijn00,

Seems to work, thanks ! 

I have questions about the Notification. To give you the context, I'm developing an app that allows the user to listen a radio stream. So there is one mediaItem with the Url. He can only stop the player (no pause&play, neither going back). So is there a way to implement actions for the others buttons (https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/ui/PlayerNotificationManager.html) ?

Also, during the stream, I want to update the Author & Title in the Notif (without stopping the play). Before version 0.5.*, it was possible with the updateNotif method. Now, I tried the following but not working
```
CrossMediaManager.Current.MediaQueue.Current.Title = mediaItem.Title;
CrossMediaManager.Current.MediaQueue.Current.Artist = mediaItem.Artist;
```

As a workaround, using "CrossMediaManager.Current.MediaQueue.Add(mediaItem);" seems to work (with ShowNavigationControls = false, otherwise, it displays the arrow, as the queue has more than 1 elem) but it's not a proper way to do that ^^

Thanks again for your quick update 🥇

**Comment by martijn00** [comment]
Ok, thanks for the explanation i get the use case. I've got to see if that is possible using the ExoPlayer extension.

**Comment by martijn00** [comment]
@valentin-debris the code you post up here should actually work. The reason it doesn't is we still use exoplayer 2.8.8 and it needs 2.9.6 to do that. Potentially i could update to that, but then everybody needs to use VS2019 with Android R8/D8.

**Comment by valentin-debris** [comment]
@martijn00 ,

Ok, I understand. Since the workaround works, no need to upgrade the project for anybody ;) But if a personally install the nuget Xam.Plugins.Android.ExoPlayer (v 2.9.6), it should works ?

**Comment by martijn00** [comment]
No, one line of code needs to be added: https://github.com/martijn00/XamarinMediaManager/commit/966813b3acd0cb863562124f33b30d87bacd111b#diff-744ab47974856b1e866d6b4a1c23f92fR112

**Comment by martijn00** [comment]
@valentin-debris in 0.5.9 iOS notifications and control should work.

**Comment by dacorp** [comment]
Regarding _ShowNavigationControls_ and _ShowPlayPauseControls_ on NotificationManager, can you add ability to show/hide **seek** controls? Applies for @valentin-debris scenario (radio stream).
Something like:
CrossMediaManager.Current.NotificationManager.ShowSeekControls = true / false;

**Comment by halchalstudio** [comment]
Well, according to my experience the following works
CrossMediaManager.Current.MediaQueue.Current.Title = MyCustomMethod.Title;
CrossMediaManager.Current.MediaQueue.Current.Artist = MyCustomMethod.Artist;

But there are 2 problems.
First, I have to place this code before the player has been started playing. So at least information are displayed.
Second, from my radio stream I fetch meta data from regular intervals. And then if I try to update this information to the notification player, it just doesn’t work. I tried placing the code between all the states of the media player (playing, paused, loading, buffering) but it will update information only if I stop the player and play it again.

What I realised that the information is set before the player even starts playing so it is set only once.

Update notifications method doesn’t do anything for me.

**Comment by TGlev** [comment]
> Well, according to my experience the following works
> CrossMediaManager.Current.MediaQueue.Current.Title = MyCustomMethod.Title;
> CrossMediaManager.Current.MediaQueue.Current.Artist = MyCustomMethod.Artist;
> 
> But there are 2 problems.
> First, I have to place this code before the player has been started playing. So at least information are displayed.
> Second, from my radio stream I fetch meta data from regular intervals. And then if I try to update this information to the notification player, it just doesn’t work. I tried placing the code between all the states of the media player (playing, paused, loading, buffering) but it will update information only if I stop the player and play it again.
> 
> What I realised that the information is set before the player even starts playing so it is set only once.
> 
> Update notifications method doesn’t do anything for me.
> 
> (Testing this plugin for Xamarin Forms Android Radio player app)

I can confirm this

**Comment by martijn00** [comment]
As said in https://github.com/martijn00/XamarinMediaManager/issues/439#issuecomment-493008196 this will be fixed in a later release. Currently we are using an older version of ExoPlayer which prevents this.

**Comment by TGlev** [comment]
@martijn00 When will you release a new version that implements the new version of exoplayer?

**Comment by martijn00** [comment]
Probably when aapt2 with d8 and r8 become more common with Xamarin. So maybe in weeks maybe in months. Depends on demand. If you want it earlier you can get a paid version from me.

**Comment by TGlev** [comment]
How do I contact you about a version like that?

**Comment by martijn00** [comment]
Email me: Martijn at baseflow dot com

**Comment by cklenk** [comment]
@martijn00 I also have a live stream radio app I use this for. I previously used the Media Notification Manager as my radio app refreshes what's playing ~10 seconds. So I too would like ability to dynamically change the media information as the live radio is playing.

**Comment by TGlev** [comment]
How do I get the  player on the lockscreen of the phone to work? In the old version I caught an event and passed it to the MediaRemoteControl. 

`MediaManager.MediaRemoteControl.RemoteControlReceived(uiEvent);`

This is now missing however and I cant find anything in the sample. Could you please elaborate on how to do this in the new version?

**Comment by martijn00** [comment]
@TGlev please open a separate issue for that.

**Comment by martijn00** [comment]
@valentin-debris can you confirm it works in the latest?

**Comment by valentin-debris** [comment]
@martijn00 ,
I have still some troubles with the notif for iOS but the problems reported here are fixed (can remove notif on Android and notif display for iOS). I will create new ticket for the new troubles now.

**Comment by alexsotir** [comment]
For the iOS lockscreen issue please check [#690 ](https://github.com/Baseflow/XamarinMediaManager/issues/690#issuecomment-607438021)

**Comment by mixernet** [comment]
Why is it not possible to update, in Android, and is not able to display in iOS, 'AlbumImageUri' in the notifications?

**Comment by orassr** [comment]
> Well, according to my experience the following works
> CrossMediaManager.Current.MediaQueue.Current.Title = MyCustomMethod.Title;
> CrossMediaManager.Current.MediaQueue.Current.Artist = MyCustomMethod.Artist;
> 
> But there are 2 problems.
> First, I have to place this code before the player has been started playing. So at least information are displayed.
> Second, from my radio stream I fetch meta data from regular intervals. And then if I try to update this information to the notification player, it just doesn’t work. I tried placing the code between all the states of the media player (playing, paused, loading, buffering) but it will update information only if I stop the player and play it again.
> 
> What I realised that the information is set before the player even starts playing so it is set only once.
> 
> Update notifications method doesn’t do anything for me.
> 
> (Testing this plugin for Xamarin Forms Android Radio player app)

Do you have any updates on title+artist getting updated? 
Thanks:)
